### PR TITLE
WT-2917 split wtperf's configuration into per-database and per-run parts

### PR DESCRIPTION
**Pull request by keithbostic**
Replacement pull request for #2971, which was merged and reverted.

### CODE REVIEW

**Comment by keithbostic** [comment]
@agorrod, this one:
```
../bench/wtperf/wtperf_opt.i:36:49: error: unknown type name ‘uint32_t’
 #define DEF_OPT_AS_UINT32(name, initval, desc)  uint32_t name;
```
turned out to be a real pain to find.

See dd43302 for the explanation and fix, but in short, the top-level Makefile builds `docs/index.html` if it doesn't exist, and compiling `doxy.c` in `dist/s_docs` won't work if you're building from a non-standard directory.

I think it's probably wrong for the top-level Makefile to create `docs/index.html`, but I replaced the C program `dist/s_docs` was calling with a Python script, hopefully that makes the problem go away forever.

**Comment by keithbostic** [comment]
autobuild: reset this please

**Comment by keithbostic** [comment]
autobuild: retest this please

**Comment by keithbostic** [comment]
@agorrod, I've fixed a few different memory leaks in wtperf.

I explicitly reproduced and fixed this one:
```
Direct leak of 248 byte(s) in 1 object(s) allocated from:
    #0 0x4a655b in __interceptor_malloc (/home/jenkins/jenkins/workspace/wiredtiger-clang-sanitizer/build_posix/bench/wtperf/wtperf+0x4a655b)
```
I did not reproduce this one:
```
+ ../../../bench/wtperf/runners/wtperf_run.sh ../../../bench/wtperf/runners/overflow-10k.wtperf 1
src/tcmalloc.cc:283] Attempt to free invalid pointer 0x4fb1fe 
```
but I found one that matches the symptom, and I believe it's fixed.

**Comment by keithbostic** [comment]
@agorrod, while I was squirreling around in wtperf today, I made some changes that aren't strictly related to the issues you saw.

You might want to review c4757d8, 1bf54a9, 92db6dc and b9b2c0a separately from this branch, and we may want to selectively take those fixes if we discard this branch.

**Comment by keithbostic** [comment]
@agorrod, 48ee5a2 is the change we discussed, to have the monitor directory follow the same pattern as the home directory.

**Comment by daveh86** [comment]
retest this please

File: bench/wtperf/smoke.sh [hunk 1]
```diff
@@ -1,4 +1,53 @@
...
+
+# Run something with a lot of random options so we test for memory leaks
+# with more than simple cases.
+args=""
+args="$args async_threads=4,"
+args="$args checkpoint_interval=10,"
+args="$args checkpoint_threads=1,"
+args="$args checkpoint_stress_rate=100,"
+args="$args compact=true,"
+args="$args compression=snappy,"
```
Comment (daveh86) [inline bench/wtperf/smoke.sh]: The previous smoke test didn't use any compression. With snappy enabled here the test is erroring when --enable-snappy isn't set within the pull req tester

**Comment by michaelcahill** [comment]
Whenever this is merged, please ensure the commit message mentions WT-2917 and not the earlier ticket.

**Comment by agorrod** [comment]
I just ran this under valgrind and there is a memory leak:

```
$ valgrind --leak-check=full ./bench/wtperf/wtperf
==10975== Memcheck, a memory error detector
==10975== Copyright (C) 2002-2013, and GNU GPL'd, by Julian Seward et al.
==10975== Using Valgrind-3.10.0 and LibVEX; rerun with -h for copyright info
==10975== Command: ./bench/wtperf/wtperf
==10975==
==10975==
==10975== HEAP SUMMARY:
==10975==     in use at exit: 21 bytes in 1 blocks
==10975==   total heap usage: 3,914 allocs, 3,913 frees, 2,017,553 bytes allocated
==10975==
==10975== 21 bytes in 1 blocks are definitely lost in loss record 1 of 1
==10975==    at 0x4C2A944: calloc (in /usr/lib64/valgrind/vgpreload_memcheck-amd64-linux.so)
==10975==    by 0x40E5E5: dcalloc (misc.c:177)
==10975==    by 0x40ACCC: create_uris (wtperf.c:1916)
==10975==    by 0x40B9B0: start_run (wtperf.c:2224)
==10975==    by 0x40B594: start_all_runs (wtperf.c:2132)
==10975==    by 0x40CC04: main (wtperf.c:2616)
==10975==
==10975== LEAK SUMMARY:
==10975==    definitely lost: 21 bytes in 1 blocks
==10975==    indirectly lost: 0 bytes in 0 blocks
==10975==      possibly lost: 0 bytes in 0 blocks
==10975==    still reachable: 0 bytes in 0 blocks
==10975==         suppressed: 0 bytes in 0 blocks
==10975==
==10975== For counts of detected and suppressed errors, rerun with: -v
==10975== ERROR SUMMARY: 1 errors from 1 contexts (suppressed: 1 from 1)
```

I didn't try non-default configurations, but it's probably worthwhile given how much structure is changed in the pull request.

**Comment by keithbostic** [comment]
The changes in commit 5975be6 allocated `CONFIG.log_table_uri`, but never freed that memory.

**Comment by agorrod** [comment]
@keithbostic I'm surprised - that leak was fixed in: cd933b92a78ac8992d41829064ddf22cce7ec9d7, looks like that was missed in the merge. Your change has benefits over the original fix - thanks.

**Pull request merged**